server/ai_mediaserver.go

```diff
@@ -687,6 +697,154 @@ func (ls *LivepeerServer) GetLiveVideoToVideoStatus() http.Handler {
 	})
 }
 
+func (ls *LivepeerServer) CreateWhip(server *media.WHIPServer) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// two main sequential parts here
+		//
+		// WHIP : stream setup, await tracks
+		// Others: auth, selection, job setup
+		// Run them in parallel
+
+		corsHeaders(w, r.Method)
+
+		ctx := r.Context()
+		requestID := string(core.RandomManifestID())
+		ctx = clog.AddVal(ctx, "request_id", requestID)
+		streamName := r.PathValue("stream")
+		if streamName == "" {
+			http.Error(w, "Missing stream name", http.StatusBadRequest)
+			return
+		}
+		ctx = clog.AddVal(ctx, "stream", streamName)
+
+		ssr := media.NewSwitchableSegmentReader()
+
+		// prevent context cancellation error in selection due to whip completing first
+		// so use this to block whip request until selection is complete TODO improve
+		selectionComplete := make(chan bool, 1)
+
+		whipConn := media.NewWHIPConnection()
+
+		go func() {
+			internalOutputHost := os.Getenv("LIVE_AI_PLAYBACK_HOST") // TODO proper cli arg
+			if internalOutputHost == "" {
+				internalOutputHost = "rtmp://localhost/"
+			}
+			mediamtxOutputURL := internalOutputHost + streamName + "-out"
+			outputURL := ""
+			streamID := ""
+			pipelineID := ""
+			pipeline := ""
+			pipelineParams := make(map[string]interface{})
+			sourceTypeStr := "livepeer-whip"
+			queryParams := r.URL.Query().Encode()
+
+			ctx = clog.AddVal(ctx, "source_type", sourceTypeStr)
+
+			if LiveAIAuthWebhookURL != nil {
+				authResp, err := authenticateAIStream(LiveAIAuthWebhookURL, ls.liveAIAuthApiKey, AIAuthRequest{
+					Stream:      streamName,
+					Type:        sourceTypeStr,
+					QueryParams: queryParams,
+					GatewayHost: ls.LivepeerNode.GatewayHost,
+				})
+				if err != nil {
+					whipConn.Close()
+					clog.Errorf(ctx, "Live AI auth failed: %s", err.Error())
+					http.Error(w, "Forbidden", http.StatusForbidden)
+					return
+				}
+
+				if authResp.RTMPOutputURL != "" {
+					outputURL = authResp.RTMPOutputURL
+				}
+
+				if authResp.Pipeline != "" {
+					pipeline = authResp.Pipeline
+				}
+
+				if len(authResp.paramsMap) > 0 {
```

Comment:
I see there is a lot of code copied from `StartLiveVideo()`, I'd consider refactoring it and re-use. It may be okeyish to just copy-paste the code like you did, but I think there is some functionality missing (e.g. I don't see sending `stream_trace`) into Kafka, which may make it more difficult to analyze.

Reply:
Can add the stream_trace (I think that was added after we wrote this code) but wanted to avoid modifying existing ingest code paths - they are too different and the risk was too high of breaking something

Reply:
Added stream_trace and some other missing metrics in https://github.com/livepeer/go-livepeer/pull/3432/commits/7236f832e9ca979a8c2aa9e76cc28aa2c35ad6ca